barretenberg/cpp/pil/avm/avm_alu.pil

```diff
@@ -264,4 +264,141 @@ namespace avm_alu(256);
+    // Given the inputs x, y and q where x & y are inputs and q is the boolean result to the query (x > y). Then there are two scenarios
+    // (x > y) -> x - y - 1 = result, where 0 <= result < p. In our case result should not overflow the field.
+    // OR 
+    // !(x > y) -> !(x - y - 1) = (p - 1) - (x - y -1) = y - x = result, where the same applies as above.
+    // These conditions can be combined with the query, q (that x > y) as follows:
+    // (x - y - 1) * q + (y - x) (1 - q) = result
+    pol commit input_ia;
+    pol commit input_ib;
+    pol commit borrow;
+
+    //If LT, then swap ia and ib else keep the same
+    input_ia  = lt_sel * ib + lte_sel * ia;
+    input_ib  = lt_sel * ia + lte_sel * ib;
+
+    pol commit a_lo;
+    pol commit a_hi;
+    // Check input_ia is well formed from its lo and hi limbs
+    input_ia = (a_lo + 2 ** 128 * a_hi) * cmp_sel;
+
+    pol commit b_lo;
+    pol commit b_hi;
+    // Check input_ib is well formed from its lo and hi limbs
+    input_ib = (b_lo + 2 ** 128 * b_hi) * cmp_sel;
+
+    pol commit p_sub_a_lo; // p_lo - a_lo
+    pol commit p_sub_a_hi; // p_hi - a_hi
+    pol commit p_a_borrow;
+
+    // Check that (p > a) by checking (p_lo > a_lo && p_hi >= ahi) || (p_lo <= a_lo && p_hi > a_hi)
+    // First condition is if borrow = 0, second condition is if borrow = 1;
+    // (p - 1) lower 128 bits = 53438638232309528389504892708671455232
+    // (p - 1) upper 128 bits = 64323764613183177041862057485226039389 
+    p_sub_a_lo = (53438638232309528389504892708671455232 - a_lo + p_a_borrow * 2 ** 128) * cmp_sel;
+    p_sub_a_hi = (64323764613183177041862057485226039389 - a_hi - p_a_borrow) * cmp_sel;
+
+    pol commit p_sub_b_lo;
+    pol commit p_sub_b_hi;
+    pol commit p_b_borrow;
+
+    // Check that (p > b) by checking (p_lo > b_lo && p_hi >= bhi) || (p_lo <= b_lo && p_hi > b_hi)
+    // First condition is if borrow = 0, second condition is if borrow = 1;
+    p_sub_b_lo = (53438638232309528389504892708671455232 - b_lo + p_b_borrow * 2 ** 128) * cmp_sel;
+    p_sub_b_hi = (64323764613183177041862057485226039389 - b_hi - p_b_borrow) * cmp_sel;
+
+
+    // Calculate the combined relation: (a - b - 1) * q + (b -a ) * (1-q)
+    // Check that (a > b) by checking (a_lo > b_lo && a_hi >= bhi) || (alo <= b_lo && a_hi > b_hi)
+    // First condition is if borrow = 0, second condition is if borrow = 1;
+    pol A_SUB_B_LO = a_lo - b_lo - 1 + borrow * 2 ** 128;
+    pol A_SUB_B_HI = a_hi - b_hi - borrow;
+
+    // Check that (a <= b) by checking (b_lo >= a_lo && b_hi >= a_hi) || (b_lo > a_lo && b_hi >= a_hi)
+    // First condition is if borrow = 0, second condition is if borrow = 1;
+    pol B_SUB_A_LO = b_lo - a_lo + borrow * 2 ** 128;
+    pol B_SUB_A_HI = b_hi - a_hi - borrow;
+
+    pol commit res_lo;
+    pol commit res_hi;
+
+    // If this is a LT operation, we already swapped the inputs so the result of ic is still correct
```

Comment:
So I believe this is fine for the following reasons. The GT check later on is the check that you are suggesting.

1.  If x >=y but a sequencer invalidly claims LT(x, y, 1), i.e. that x < y.
2. The circuit correctly inverts such that we test GT(y, x, 1), i.e. that y > x.
3. `IS_GT` is set to 1 since we are performing an LT check on already swapped inputs.
4. the res_lo and res_hi will select `A_SUB_B_LO` and `A_SUB_B_HI` respectively. In our case it would be:
   a) res_lo = y_lo - x_lo -1 + borrow * 2**128
   b) res_hi = y_hi - x_hi - borrow.
   
5. This should fail when it comes to the range check because either
  a) x == y && borrow = 1, causes res_hi to underflow, or
  b) x == y && borrow = 0,causes res_lo to underflow
  c) x > y   && borrow = 1, causes res_hi to underflow
  d) x > y   && borrow = 0, either res_lo & res_hi underflows depending on input
  
 6) The only way to pass this is for the sequencer to validly claim LT(x,y,0), i.e. x >= y.  This should also be what you are suggesting by `But if the inputs for LT are such that x >= y, I think we should swap them and prove LTE and invert the result.`. This happens automatically through the checks below

```
pol IS_GT = op_lt * ic + (1 - ic) * op_lte;

res_lo = (A_SUB_B_LO * IS_GT + B_SUB_A_LO * (1 - IS_GT)) * cmp_sel;

res_hi = (A_SUB_B_HI * IS_GT + B_SUB_A_HI * (1 - IS_GT)) * cmp_sel;
```